### PR TITLE
Add jenkins for deploys

### PR DESCRIPTION
**Pull request by gyanesh-mishra**
Migrated Jenkins over from the Digital Marketplace challenge.
Jenkins configuration needs to happen per project on GUI differently depending on the needs of the project.

Not sure if we need to merge this branch because not all projects need it and it still would require configuration and just keep it alive as a PR/Branch 🤷‍♂ 

Looking for feedback/questions/concerns.

Part of deploying this, we also need two servers (lightsail instances are recommended) and some configuration around those as well. I dont think that's relevant to the documentation because it has nothing to do with deployments or Jenkins. I can try to make the instructions as verbose as possible but if someone who has never done Jenkins/deployments picks it up and somehow manages to deploy following hand held instructions. They would have no clue what to do at the sign of first error

### CODE REVIEW

File: deployment/jenkins/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
+FROM jenkins/jenkins:2.190.1
+
+USER root
```
Comment (fwkendall) [inline deployment/jenkins/Dockerfile:3]: Whoa we need root here? :o

File: deployment/jenkins/jobs/Digital Mktplace Library/config.xml [hunk 2]
```diff
@@ -0,0 +1,70 @@
+<?xml version='1.1' encoding='UTF-8'?>
```
Comment (fwkendall) [inline deployment/jenkins/jobs/Digital Mktplace Library/config.xml:1]: should we rename this file?

File: deployment/jenkins/jobs/Digital Mktplace Library/state.xml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<?xml version='1.1' encoding='UTF-8'?>
```
Comment (fwkendall) [inline deployment/jenkins/jobs/Digital Mktplace Library/state.xml:1]: This file name, too

File: deployment/jenkins/jobs/Digital Mktplace Library/config.xml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+<org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject plugin="workflow-multibranch@2.21">
+  <actions/>
+  <description>Code Challenge for the Digital Marketplace Sprint With Us opportunity (Oct 2019)</description>
```
Comment (fwkendall) [inline deployment/jenkins/jobs/Digital Mktplace Library/config.xml:4]: we should clean up any references to digital marketplace

File: deployment/scripts/deploy-application.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+set -exv -o pipefail
```
Comment (fwkendall) [inline deployment/scripts/deploy-application.sh:2]: 👍

File: docker-compose-cd.yml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    image: database:${GIT_COMMIT:-latest}
+    container_name: database
+    build:
+      context: database
+    ports:
+      - 5432:5432
+    environment:
+      - POSTGRES_USER=${POSTGRES_USER:-admin}
+      - POSTGRES_PASSWORD=${POSTGRES_PASSWORD:-password}
+      - POSTGRES_DB=dm_library
```
Comment (fwkendall) [inline docker-compose-cd.yml]: is this "library" naming specific to the last SWU, too?

Comment (fwkendall) [reply docker-compose-cd.yml]: oh yeah, guessing dm = Digital Marketplace here

File: docker-compose-cd.yml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+  ####### Server Database Migration Definition ###########
+
+  server-migrate:
+    restart: on-failure
+    image: server-migrate:${GIT_COMMIT:-latest}
+    container_name: server-migrate
+    build:
+      context: server
+    command: python manage.py create-db
```
Comment (fwkendall) [inline docker-compose-cd.yml:69]: so this is what we would replace with a migration script instead?

Comment (fwkendall) [reply docker-compose-cd.yml:69]: but we would still need to call create-db somewhere, also

**Review by fwkendall — CHANGES_REQUESTED** [review]
Looks great! Main thing from me is that I'd like to make sure all library / digital marketplace references are removed 😄

File: deployment/jenkins/Dockerfile (hunk 1 above)
Comment (gyanesh-mishra) [reply deployment/jenkins/Dockerfile:3]: Yeah any apt-get commands need to run as sudo, so root is needed :(

File: deployment/jenkins/jobs/Digital Mktplace Library/config.xml (hunk 2 above)
Comment (gyanesh-mishra) [reply deployment/jenkins/jobs/Digital Mktplace Library/config.xml:1]: It'll all have to be done manually using jenkins UI, I can try to generalize, but it won't matter, and I may mess up some config accidentally

File: docker-compose-cd.yml (hunk 6 above)
Comment (gyanesh-mishra) [reply docker-compose-cd.yml]: Yes! good catch, updating

File: docker-compose-cd.yml (hunk 7 above)
Comment (gyanesh-mishra) [reply docker-compose-cd.yml:69]: yeah, given current stack this was my approach 😅

File: deployment/jenkins/Dockerfile (hunk 1 above)
Comment (fwkendall) [reply deployment/jenkins/Dockerfile:3]: Gross!

**Comment by fwkendall** [comment]
Logged WEB-89 to track the leftover renaming/sanitization changes :) Looks good to go!

**Pull request merged**